src/middlewared/middlewared/plugins/activedirectory.py

```diff
@@ -0,0 +1,1147 @@
+            await self.stop()
+        if start:
+            await self.start()
+
+        if not stop and not start and new['enable']:
+            await self.middleware.call('service.restart', 'cifs')
+
+        return await self.config()
+
+    @private
+    async def _set_state(self, state):
+        await self.middleware.call('cache.put', 'AD_State', state.name)
+
+    @private
+    async def get_state(self):
+        """
+        Check the state of the AD Directory Service.
+        See DSStatus for definitions of return values.
+        :DISABLED: Service is not enabled.
+        If for some reason, the cache entry indicating Directory Service state
+        does not exist, re-run a status check to generate a key, then return it.
+        """
+        ad = await self.config()
+        if not ad['enable']:
+            return 'DISABLED'
+        else:
+            try:
+                return (await self.middleware.call('cache.get', 'AD_State'))
+            except KeyError:
+                await self.started()
+                return (await self.middleware.call('cache.get', 'AD_State'))
+
+    @private
+    async def start(self):
+        """
+        Start AD service.
+        """
+        ad = await self.config()
+        smb = await self.middleware.call('smb.config')
+        state = await self.get_state()
+        if state in [DSStatus['JOINING'], DSStatus['LEAVING']]:
+            raise CallError(f'Active Directory Service has status of [{state.value}]. Wait until operation completes.', errno.EBUSY)
+
+        await self._set_state(DSStatus['JOINING'])
+        await self.middleware.call('datastore.update', self._config.datastore, ad['id'], {'ad_enable': True})
+        await self.middleware.call('etc.generate', 'hostname')
+
+        """
+        Kerberos realm field must be populated so that we can perform a kinit
+        and use the kerberos ticket to execute 'net ads' commands.
+        """
+
+        if not ad['kerberos_realm']:
+            realms = await self.middleware.call('kerberos.realm.query', [('realm', '=', 'DS_TYPE_ACTIVEDIRECTORY')])
+
+            if realms:
+                await self.middleware.call('datastore.update', self._config.datastore, ad['id'], {'ad_kerberos_realm': realms[0]['id']})
+            else:
+                await self.middleware.call('datastore.insert', 'directoryservice.kerberosrealm', {'krb_realm': ad['domainname'].upper()})
+
+        await self.middleware.call('kerberos.start')
+
+        """
+        'workgroup' is the 'pre-Windows 2000 domain name'. It must be set to the nETBIOSName value in Active Directory.
+        This must be properly configured in order for Samba to work correctly as an AD member server.
+        'site' is the ad site of which the NAS is a member. If sites and subnets are unconfigured this will
+        default to 'Default-First-Site-Name'.
+        """
+
+        if not ad['site']:
+            new_site = await asyncio.wait_for(self.get_site(), 10)
```

Comment:
This may raise an exception on timeout, is this handled/expected?

Reply:
Added some handling here.